### PR TITLE
Refactor DigitalOcean modules

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Fix adds refactoring of DigitalOcean modules by adding
common functionality to module_utils. Future module development can use the common code.

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/digital_ocean.py
lib/ansible/modules/cloud/digital_ocean/digital_ocean_block_storage.py
lib/ansible/modules/cloud/digital_ocean/digital_ocean_tag.py

##### ANSIBLE VERSION
```
2.4devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
@BondAnthony Could you please review this ?

**Comment by Akasurde** [comment]
ready_for_review

File: lib/ansible/module_utils/digital_ocean.py
Comment (BondAnthony) [inline lib/ansible/module_utils/digital_ocean.py]: You will need to include the OAUTH_TOKEN environment variable also.

File: lib/ansible/module_utils/digital_ocean.py
Comment (BondAnthony) [inline lib/ansible/module_utils/digital_ocean.py:32]: Do we really want to use `_text` while it's still marked as unstable. This could potentially have an impact a number of DO modules if we begin to use this before it's stable.

**Review by BondAnthony — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/digital_ocean.py
Comment (abadger) [reply lib/ansible/module_utils/digital_ocean.py:32]: We do.  I'm going to mark it stable for either 2.4 or 2.5.  The API has changed in only one subtle way in the past few releases (the default error handler changed from surrogate_or_replace to surrogate_then_replace as I found one cornercase where surrogate_or_replace could traceback).

the to_* methods make encoding issues and conversion from bytes and text much simpler than trying to ad hoc a solution.  And as a bonus, as long as it's marked unstable, I've been reviewing the code that uses it to make sure I fix them if I make any changes that would break (so far I've been lucky on that front.  the only things I've had to change were when I first introduced the functions and only a few things were using them).  Once I declare the interface stable I'll have to create new versions of the functions if I make backwards incompatible changes and then I'll rely more heavily on maintainers to port to new versions on a deprecation cycle.

Comment (BondAnthony) [reply lib/ansible/module_utils/digital_ocean.py:32]: @abadger thank you for the explanation.

**Review by BondAnthony — APPROVED** [review]

**Comment by BondAnthony** [comment]
shipit

**Comment by Akasurde** [comment]
@BondAnthony Thanks for review.

**Comment by BondAnthony** [comment]
@abadger could you please review this when you have a few minutes. Once merged we can move forward with a couple other module PR.

**Comment by abadger** [comment]
@Akasurde Oh... one thing about this... the module_utils is licensed GPLv3+... we like to have module_utils licensed BSD.... Is that code you have permission to relicense or is it code from a random asortment of contributors (coming out of the modules themselves)?

**Comment by Akasurde** [comment]
@abadger @harneksidhu @mattclay @gundalow @kontrafiktion 

I am refactoring Digital Ocean modules by moving module level code into module_utils. This requires change in licensing of code i.e. from `GPLv3+` to `BSD`

(the modules stays GPLv3+ but the module_utils will be BSD)

I would you like to have your permission for this change. Could you please signoff this change ?

Thanks,
Abhijeet

**Comment by abadger** [comment]
I give permission to relicense the pieces of my contributions to the digital ocean modules which are being moved to module_utils from GPLv3+ to BSD.

Note:  The version of the BSD license that we would be licensing under can be seen here: https://github.com/ansible/ansible/blob/b26ee657b56bb9195f07c6920de08a9ee4ddbeac/lib/ansible/module_utils/basic.py#L11

**Comment by Akasurde** [comment]
I give permission to relicense the pieces of my contributions to the digital ocean modules which are being moved to module_utils from GPLv3+ to BSD.

**Comment by mattclay** [comment]
I give permission to relicense the pieces of my contributions to the digital ocean modules which are being moved to module_utils from GPLv3+ to BSD.

**Comment by gundalow** [comment]
I give permission to relicense the pieces of my contributions to the digital ocean modules which are being moved to module_utils from GPLv3+ to BSD.

**Comment by Akasurde** [comment]
@harneksidhu @kontrafiktion Could you please provide permission for re-licensing as mentioned above ?

**Comment by Akasurde** [comment]
@harneksidhu @kontrafiktion Could you please provide permission for re-licensing as mentioned above ?

**Comment by kontrafiktion** [comment]
I give permission to relicense the pieces of my contributions to the digital ocean modules which are being moved to module_utils from GPLv3+ to BSD.

**Comment by Akasurde** [comment]
@harneksidhu Could you please provide permission for re-licensing as mentioned above ?

**Comment by Akasurde** [comment]
@abadger Is there anything like auto-permission or something with which we can proceed with merge of this PR? I am concerned as other digital_ocean modules are pending due to this PR ?

**Comment by abadger** [comment]
I don't believe so.  Our options would be to reimplement the parts that
cannot be relicensed or allow gplv3+ module_utils

**Comment by BondAnthony** [comment]
Well, it looks like reimplementation would be the best option instead of adding GPLv3 to module_utils code. I would like to break down some of the digital_ocean module. We have duplication in functionality and a large dependency on dopy which no longer supported. Thoughts?

**Comment by harneksidhu** [comment]
I give permission to relicense the pieces of my contributions to the digital ocean modules which are being moved to module_utils from GPLv3+ to BSD.

**Comment by Akasurde** [comment]
@harneksidhu @abadger @mattclay @gundalow @kontrafiktion Thanks for your permission. 

@abadger Could you please merge this ? Thanks in advance.

**Comment by abadger** [comment]
Merged to devel for the 2.4 release.  Yay!  Congratulations!

**Comment by Akasurde** [comment]
@abadger Thanks for merge. Thanks everyone for their permission and contributions.

**Pull request merged**
